### PR TITLE
modified SE method

### PR DESCRIPTION
**Pull request by stemangiola**

### CODE REVIEW

**Comment by stemangiola** [comment]
Hello @mblue9, I have sped up the method aggregate method for SE. See if the SE TCGA is fast enough.

As a workflow, I would abandon the pure tibble, in favour of SE. For example:

1. get SE from TCGA with unique transcript names as a feature.
2. use aggregate duplicates (e.g. on gene symbol)
3. when we display stats and visualise data, we can easily use the column gene symbol, even though the row names of the SE are the collapsed transcript names

Basically, with tidy, we don't have to worry anymore about replacing one name with the other.

Having said that, I will try to implement the fast aggregate based on tibble in the next few days. Of course, if you want to have a crack yourself that would also be good.

**Comment by mblue9** [comment]
Great, I was about to change my workflow to try using TCGABiolinks to import the data as SE so this could work well.

**Comment by stemangiola** [comment]
> Great, I was about to change my workflow to try using TCGABiolinks to import the data as SE so this could work well.

With SE there are two option

1. you do your analyses on transcripts/isoforms/ensembl and visualise/plot gene symbol
2. aggregate based on gene symbol and do your analysis normally, and still display and plot with gene symbol column

**Comment by stemangiola** [comment]
Alright,

I created a more efficient implementation for big datasets. This processes your 2BILLION record data frame in 5 minutes, rather than eternity.

it takes 30 seconds to do everything, but 4 minutes to return the datasets as R has to duplicate the whole dataset (30Gb ) in RAM.

I would suggest any way to not work with all 97 annotation columns, but communicate that when working with huge datasets, to decide what is the reasonable subset of annotation you need.

Future general solution:
- for tibble use data.table, but it does not work in our setup for some probable dependency fight
- migrate everything to SE as mentioned, after having made everything more efficient.

In any case, a LOT of work :)

**Comment by mblue9** [comment]
>I created a more efficient implementation for big datasets. This processes your 2BILLION record data frame in 3 minutes, rather than eternity.

>it takes 30 seconds to do everything, but 2.5 minutes to return the datasets as R has to duplicate the whole dataset (30Gb ) in RAM.

>I would suggest any way to not work with all 97 annotation columns, but communicate that when working with huge datasets, to decide what is the reasonable subset of annotation you need.

What code are you testing with? 
The counts_se dataset imported with tcgabiolinks is 24 million rows not 2 billion and I don't think I have a choice on what columns get imported into the counts_se. Then select isn't currently working on the size of the dataset so I can't reduce the annotation.

**Comment by stemangiola** [comment]
I mean 24 million times 97 columns. Is select not working for tibble?

**Comment by stemangiola** [comment]
I see I will test tomorrow. Can you please pass me the data frame text file
to reproduce the error?

**Comment by mblue9** [comment]
Will do, thanks for having a look!

**Comment by stemangiola** [comment]
> Will do, thanks for having a look!

For records this is the link

https://unimelbcloud-my.sharepoint.com/:u:/g/personal/maria_doyle_unimelb_edu_au/EcI-F26LLu9NuWoOcgpiATsBLci49UwXrED2A_voIxdg5Q?e=hMcE8j

**Comment by stemangiola** [comment]
It works for me, maybe try again to pull this branch

```r
> counts_tt 
# A tibble: 8,663,925 x 5
   Hugo_Symbol  Entrez_Gene_Id name            value her2_group
   <chr>        <chr>          <chr>           <dbl> <chr>     
 1 LOC100130426 100130426      TCGA-5T-A9QA-01     0 HER2_low  
 2 LOC100130426 100130426      TCGA-A1-A0SE-01     0 HER2_low  
 3 LOC100130426 100130426      TCGA-A1-A0SH-01     0 HER2_low  
 4 LOC100130426 100130426      TCGA-A1-A0SJ-01     0 HER2_low  
 5 LOC100130426 100130426      TCGA-A1-A0SK-01     0 HER2_neg  
 6 LOC100130426 100130426      TCGA-A1-A0SM-01     0 HER2_pos  
 7 LOC100130426 100130426      TCGA-A1-A0SP-01     0 HER2_neg  
 8 LOC100130426 100130426      TCGA-A2-A04Q-01     0 HER2_low  
 9 LOC100130426 100130426      TCGA-A2-A04R-01     0 HER2_low  
10 LOC100130426 100130426      TCGA-A2-A04T-01     0 HER2_low  
# … with 8,663,915 more rows
> counts_dupsrem <- counts_tt %>% aggregate_duplicates()
tidybulk says: for big data sets (>100 samples) this efficient implementation aggregates count columns and keeps the first instance for sample and transcript annotations
> counts_dupsrem 
# A tibble: 8,660,706 x 5
   Hugo_Symbol  Entrez_Gene_Id name            value her2_group
   <chr>        <chr>          <chr>           <dbl> <chr>     
 1 LOC100130426 100130426      TCGA-5T-A9QA-01     0 HER2_low  
 2 LOC100130426 100130426      TCGA-A1-A0SE-01     0 HER2_low  
 3 LOC100130426 100130426      TCGA-A1-A0SH-01     0 HER2_low  
 4 LOC100130426 100130426      TCGA-A1-A0SJ-01     0 HER2_low  
 5 LOC100130426 100130426      TCGA-A1-A0SK-01     0 HER2_neg  
 6 LOC100130426 100130426      TCGA-A1-A0SM-01     0 HER2_pos  
 7 LOC100130426 100130426      TCGA-A1-A0SP-01     0 HER2_neg  
 8 LOC100130426 100130426      TCGA-A2-A04Q-01     0 HER2_low  
 9 LOC100130426 100130426      TCGA-A2-A04R-01     0 HER2_low  
10 LOC100130426 100130426      TCGA-A2-A04T-01     0 HER2_low  
# … with 8,660,696 more rows
```

**Comment by mblue9** [comment]
Pulled again and aggegate_duplicates works, thanks.
Now just having an issue with pivot_sample dropping the her2_group column so I can't e.g  colour by group
```
counts_tt %>%
aggregate_duplicates() %>% 
keep_abundant(factor_of_interest = her2_group) %>% 
reduce_dimensions(method = "PCA") %>% 
pivot_sample()

# A tibble: 422 x 4
   name            .abundant    PC1    PC2
   <chr>           <lgl>      <dbl>  <dbl>
 1 TCGA-5T-A9QA-01 TRUE      -7.38  12.6  
 2 TCGA-A1-A0SE-01 TRUE      -3.15  -8.93 
 3 TCGA-A1-A0SH-01 TRUE       0.374 -1.22 
 4 TCGA-A1-A0SJ-01 TRUE      -1.34  -9.15 
 5 TCGA-A1-A0SK-01 TRUE      10.4    6.02 
 6 TCGA-A1-A0SM-01 TRUE      -3.25   0.191
 7 TCGA-A1-A0SP-01 TRUE      15.8   -0.269
 8 TCGA-A2-A04Q-01 TRUE      15.3   -0.728
 9 TCGA-A2-A04R-01 TRUE      -8.15   1.29 
10 TCGA-A2-A04T-01 TRUE      20.5    3.74 
```

**Comment by stemangiola** [comment]
How many different values of her2 are there before pivot_sample?

**Comment by mblue9** [comment]
3
```
> counts_PCA %>% distinct(her2_group)
# A tibble: 3 x 1
  her2_group
  <chr>     
1 HER2_low  
2 HER2_neg  
3 HER2_pos  
```
